File: backend/src/main/java/com/crm/mapper/BenutzerMapper.java
package com.crm.mapper;

import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.security.crypto.password.PasswordEncoder;

import com.crm.dto.BenutzerCreateDTO;
import com.crm.dto.BenutzerDTO;
import com.crm.entity.Benutzer;
import com.crm.entity.enums.BenutzerRolle;

public final class BenutzerMapper {

    private BenutzerMapper() {}

    public static BenutzerDTO toDTO(Benutzer e) {
        return new BenutzerDTO(
                e.getId(),
                e.getBenutzername(),
                e.getVorname(),
                e.getNachname(),
                e.getEmail(),
                e.getRollen().stream().map(Enum::name).toList(),
                e.isAktiv(),
                e.getCreatedAt(),
                e.getUpdatedAt()
        );
    }

    public static Benutzer toEntity(BenutzerCreateDTO dto, PasswordEncoder encoder) {
        Benutzer e = new Benutzer();
        e.setBenutzername(dto.benutzername());
        e.setPasswort(encoder.encode(dto.passwort()));
        e.setVorname(dto.vorname());
        e.setNachname(dto.nachname());
        e.setEmail(dto.email());
        e.setRollen(parseRollen(dto.rollen()));
        e.setAktiv(dto.aktiv());
        return e;
    }

    public static void applyToEntity(BenutzerCreateDTO dto, Benutzer e, PasswordEncoder encoder) {
        e.setVorname(dto.vorname());
        e.setNachname(dto.nachname());
        e.setEmail(dto.email());
        e.setRollen(parseRollen(dto.rollen()));
        e.setAktiv(dto.aktiv());
        if (dto.passwort() != null && !dto.passwort().isBlank()) {
            e.setPasswort(encoder.encode(dto.passwort()));
        }
    }

    private static Set<BenutzerRolle> parseRollen(Set<String> rollen) {
        return rollen.stream()
                .map(BenutzerRolle::valueOf)
                .collect(Collectors.toSet());
    }
}

File: ciam/src/main/kotlin/com/crm/ciam/mapper/BenutzerMapper.kt
package com.crm.ciam.mapper

import com.crm.ciam.dto.BenutzerCreateDTO
import com.crm.ciam.dto.BenutzerDTO
import com.crm.ciam.entity.Benutzer
import com.crm.ciam.entity.enums.BenutzerRolle
import org.springframework.security.crypto.password.PasswordEncoder

object BenutzerMapper {

    fun toDTO(e: Benutzer): BenutzerDTO = BenutzerDTO(
        id = e.id,
        benutzername = e.benutzername,
        vorname = e.vorname,
        nachname = e.nachname,
        email = e.email,
        rollen = e.rollen.map { it.name },
        aktiv = e.aktiv,
        createdAt = e.createdAt,
        updatedAt = e.updatedAt
    )

    fun toEntity(dto: BenutzerCreateDTO, encoder: PasswordEncoder): Benutzer {
        val e = Benutzer()
        e.benutzername = dto.benutzername
        e.passwort = encoder.encode(dto.passwort)
        e.vorname = dto.vorname
        e.nachname = dto.nachname
        e.email = dto.email
        e.rollen = parseRollen(dto.rollen)
        e.aktiv = dto.aktiv
        return e
    }

    fun applyToEntity(dto: BenutzerCreateDTO, e: Benutzer, encoder: PasswordEncoder) {
        e.vorname = dto.vorname
        e.nachname = dto.nachname
        e.email = dto.email
        e.rollen = parseRollen(dto.rollen)
        e.aktiv = dto.aktiv
        if (!dto.passwort.isNullOrBlank()) {
            e.passwort = encoder.encode(dto.passwort)
        }
    }

    private fun parseRollen(rollen: Set<String>): MutableSet<BenutzerRolle> =
        rollen.map { BenutzerRolle.valueOf(it) }.toMutableSet()
}
